Add select--stroke modifier and change to default borderless select

adrianababakanian opened the pull request:
Closes https://github.com/mapbox/assembly/issues/982

This PR adds the ~`select--border-0` modifier class to support borderless select elements out-of-the-box. It removes the `select`'s default box-shadow and left padding.~ `select--stroke` modifier and updates select elements to be unstroked by default.

@samanpwbb for review -- I'm curious if you think the padding right of the `select-arrow` should also be reduced or removed, in which case there would be some additional work to be done here.

adrianababakanian committed:
Add select--border-0 class

adrianababakanian committed:
Update padding and tests

samanpwbb reviewed (commented):
@adrianababakanian this looks like a solid start! 

> I'm curious if you think the padding right of the `select-arrow` should also be reduced or removed

Looking to the figma files for guidance, I think we do want a small amount of right padding yes!

<img width="625" alt="Screen Shot 2021-04-12 at 9 24 36 AM" src="https://user-images.githubusercontent.com/108094/114428345-ee349d00-9b70-11eb-8f30-e04914877155.png">

Before I approve or reject this, I have a couple questions: 

1. Did you consider the trade-offs of making the _stroke_ a modifier and the unstroked style the new base style? That might be more in keeping with Assembly's additive styling model, but at the cost of a change that might be considered "breaking" (although that's up for debate). 

2. Did you consider styling disabled/active states differently for unstroked selects? I wonder, for example, if we should only apply background-color changes to stroked selects.

adrianababakanian commented:
Thanks for the feedback @samanpwbb !

> Looking to the figma files for guidance, I think we do want a small amount of right padding yes!-

👍 

> Did you consider the trade-offs of making the stroke a modifier and the unstroked style the new base style? 

I didn't consider this, though I do see how making the unstroked style the default would be more in keeping with the additive styling model that you mentioned. Should we move forward with switching the modifier to something like `select--stroke`, similar to `btn--stroke`?

> Did you consider styling disabled/active states differently for unstroked selects?

Great callout. I do think the active state makes sense to keep the same as the stroked select -- i.e., the arrow color becomes darker and font color remains the same.

<img width="651" alt="Screen Shot 2021-04-13 at 9 52 09 AM" src="https://user-images.githubusercontent.com/15935667/114591255-8f3d5980-9c3e-11eb-9ae1-2c88665a7f8a.png">

But, we might want to handle the disabled state differently, as it doesn't look great with the current state of this PR given the lack of left padding:

<img width="165" alt="Screen Shot 2021-04-13 at 9 56 56 AM" src="https://user-images.githubusercontent.com/15935667/114591316-9d8b7580-9c3e-11eb-8d8a-f731caae5721.png">

I'm not seeing anything explicitly in the style guide for how to handle this case -- do you have any thoughts?

samanpwbb commented:
> Should we move forward with switching the modifier to something like `select--stroke`, similar to `btn--stroke`?

I am on board – @katydecorah would you be okay with this change? It may require some work to migrate for you.


> I'm not seeing anything explicitly in the style guide for how to handle this case -- do you have any thoughts?

Ya, maybe something closer to how we do disabled `link`? disabled-colored text and disabled-colored triangle, and leave background transparent?

katydecorah commented:
> would you be okay with this change?

👍 Yes, I think that pattern makes sense

adrianababakanian committed:
Default borderless, select--border modifier

adrianababakanian committed:
Update tests

adrianababakanian commented:
When modifying `select` to be borderless by default, I made a few modifications:

- Added a `select--stroke` modifier to add a border + additional padding to the select element.
- Changed the `select--border-${color}` modifiers to `select--${color}`. They now have the effect of coloring just the arrow when used with `select`, and both the arrow and border when used in conjunction with the `select--stroke` modifier.

<img width="765" alt="Screen Shot 2021-04-13 at 5 00 34 PM" src="https://user-images.githubusercontent.com/15935667/114636196-48218980-9c7b-11eb-9f91-e7dd2842c659.png">

@samanpwbb I wonder if it would make sense for the `select--${color}` modifiers to also change the font color for both the bordered and borderless versions. Referencing Figma, it looks like one of the bordered selects has a different font + arrow color from its border, whereas the other uses the same color for all three elements. Should the former be the default coloring, and the latter the result if a user explicitly sets a color via a modifier?

<img width="277" alt="Screen Shot 2021-04-13 at 5 11 47 PM" src="https://user-images.githubusercontent.com/15935667/114636243-65eeee80-9c7b-11eb-8e4e-0f3d53acc771.png">

samanpwbb commented:
> I wonder if it would make sense for the `select--${color}` modifiers to also change the font color for both the bordered and borderless versions. 

For unstroked, definitely.

Not sure about stroked - we could go either way.. I know @katydecorah originally modelled these after text inputs, which do not change text color.

> Referencing Figma, it looks like one of the bordered selects has a different font + arrow color from its border, whereas the other uses the same color for all three elements.

What you're seeing is the difference between the active and inactive state. Active state should use `-dark` variants for everything, while inactive state should have a light border, and default colored text and triangle. 

Also, looking at your screenshot, I wonder if we should adjust the default select style to use a darker triangle?

adrianababakanian committed:
Address feedback

adrianababakanian committed:
Fix linting + update snapshots

adrianababakanian commented:
Thanks for the clarifications and continued reviews @samanpwbb ! I updated the default arrow to use the default form color rather than the lighter default form color, so now it's the same as the font color. The active state now also uses `-dark` variants for everything. 

<img width="146" alt="Screen Shot 2021-04-14 at 9 44 46 AM" src="https://user-images.githubusercontent.com/15935667/114748483-d727b380-9d06-11eb-8836-7c0026e8ffe5.png">
<img width="147" alt="Screen Shot 2021-04-14 at 9 42 55 AM" src="https://user-images.githubusercontent.com/15935667/114748495-d8f17700-9d06-11eb-9f69-e11ebf96b832.png">

Right now the colored stroked selects use an updated font color as well. I like how it creates consistency with how unstroked selects handle color variants, but at the same time I see the value in keeping them modeled after text inputs (which don't change colors). @katydecorah I'd appreciate any thoughts you might have on this!


<img width="770" alt="Screen Shot 2021-04-14 at 9 42 29 AM" src="https://user-images.githubusercontent.com/15935667/114748476-d4c55980-9d06-11eb-815c-2730bb65eb6f.png">

samanpwbb commented on src/forms.css:
```diff
@@ -42,10 +42,14 @@
 .input:focus,
 .textarea:focus,
-.select:focus {
+.select.select--stroke:focus {
```
I don't think we need the `.select` part of this selector, can be .select--stroke:focus

adrianababakanian replied:
This one is a bit interesting. When the `.select:focus { box-shadow: none }` rule is included, the `.select` part of the selector is necessary. Otherwise, a select element using a `select--stroke` modifier loses its border when focused. Here's a screenshot where the focus is currently on the first stroked element when `.select` is removed from the above selector. 

<img width="768" alt="Screen Shot 2021-04-14 at 2 24 45 PM" src="https://user-images.githubusercontent.com/15935667/114781303-59c36980-9d2d-11eb-9fca-3fb8b2cf9f96.png">


Would be happy to hear ideas for better approaches! This also might change depending on how we decided to handle the `.select:focus` rule (i.e. the comment below).

samanpwbb replied:
aha, that makes sense.

katydecorah commented:
> Right now the colored stroked selects use an updated font color as well. I like how it creates consistency with how unstroked selects handle color variants

@adrianababakanian While I love the way it looks, we should make sure that the font color for text has an [accessible contrast ratio](https://webaim.org/resources/contrastchecker/). My vote is to optimize for legibility over consistency. I know the Assembly color palette will be updated in v1, so perhaps the font color will soon have improved color contrast that would allow us to successfully update the font color without losing legibility.

samanpwbb commented on src/forms.css:
```diff
@@ -42,10 +42,14 @@
 
+.select:focus {
+  box-shadow: none;
+}
```
I think we can cut this rule.

adrianababakanian replied:
Without this rule, the default focus styling is applied (shown below). I didn't think this looked great due to the lack of left padding -- what do you think? 

<img width="772" alt="Screen Shot 2021-04-14 at 2 19 40 PM" src="https://user-images.githubusercontent.com/15935667/114780609-7dd27b00-9d2c-11eb-8d55-a9404798946a.png">

samanpwbb replied:
Aha, the more prominent tab style, even though it's ugly, might actually be the best move. Right now tab active state for selects is disorienting because it's so subtle.

Thinking about this more, what if we did this: write an even higher priority tab state selector that used `[data-assembly-focus-control='visible']`, and that applied _two box shadows_: one that is a single stroke, and a second box shadow that matches the tab active states we use elsewhere in assemby (for links, buttons, ect).

adrianababakanian replied:
> Right now tab active state for selects is disorienting because it's so subtle.

That's true, just changing the text + arrow color to be slightly darker isn't noticeable unless you know to look for it. 

> what if we did this: write an even higher priority tab state selector that used [data-assembly-focus-control='visible'], and that applied two box shadows: one that is a single stroke, and a second box shadow that matches the tab active states we use elsewhere in assemby (for links, buttons, ect).

This makes a lot of sense to me! Maybe with one caveat -- for stroked buttons, the focus styling actually uses just one box-shadow and removes the thinner, colored box-shadow applied via the relevant color modifier. What do you think about using the same approach for selects?

<img width="570" alt="Screen Shot 2021-04-14 at 2 57 23 PM" src="https://user-images.githubusercontent.com/15935667/114785615-ce000c00-9d31-11eb-90c3-107e0718df96.png">
<img width="574" alt="Screen Shot 2021-04-14 at 2 57 14 PM" src="https://user-images.githubusercontent.com/15935667/114785616-ce98a280-9d31-11eb-9fc6-919a3d84c033.png">


This does break consistency with how colored textareas handle focus, though. This would go beyond the scope of this PR, but maybe the correct approach is to update both selects and _also_ textareas to have two box-shadows when focused? To help illustrate, the yellow textarea here has a tab active state, but it's very subtle.

<img width="998" alt="Screen Shot 2021-04-14 at 2 55 08 PM" src="https://user-images.githubusercontent.com/15935667/114785425-86798000-9d31-11eb-8451-5ab0015921bf.png">

samanpwbb replied:
@adrianababakanian ooh thanks for documenting this. We should ticket making the focus state for textareas and text inputs more prominent, maybe also with the double box-shadow technique?

adrianababakanian replied:
I went ahead with adding a double box-shadow for focused stroked selects in this PR per your suggestion @samanpwbb , I'll go ahead and open a ticket for improving focus states for textareas and text inputs too 👍

samanpwbb commented on src/forms.css:
```diff
@@ -126,11 +130,11 @@
   color: var(--darken50);
   background-color: var(--disabled-interactive-color);
-  box-shadow: inset 0 0 0 1px var(--disabled-interactive-color-dark);
+  box-shadow: inset 0 0 0 1px var(--disabled-interactive-color-dark) !important;
```
Is this `!important` required?

adrianababakanian replied:
Without the `!important`, the disabled `select--stroke` elements retain their colored borders (see below). I wonder if there's a better way to achieve the desired effect without using `!important`? I thought using it might make sense given the context -- that is, we always want disabled `select--stroke`s to use the disabled color variable.

<img width="760" alt="Screen Shot 2021-04-14 at 2 28 29 PM" src="https://user-images.githubusercontent.com/15935667/114781871-22a18800-9d2e-11eb-80cb-4f0e005ecd40.png">

samanpwbb replied:
makes sense to keep !important

samanpwbb reviewed (commented):
@adrianababakanian this is coming together! Noticed a couple minor issues.

Another thing I noticed, that is also present on previous version: we don't have good `[data-assembly-focus-control='visible']` styling for selects. Is this something we could fix? Notice how other form types take on a prominent `:focus` style, but selects do not.

adrianababakanian commented:
> My vote is to optimize for legibility over consistency.

Really great point @katydecorah ! I think that makes sense. Optimizing for legibility, I see three potential paths forward:

1. Use the default font color for both stroked and unstroked selects regardless of the color modifier. This means that applying color variants to the default selects only affects the arrow color, which does look a little funny.
2. Use the `-medium` variants for each color from the new style guide as the text + arrow colors when color modifiers are applied (for both stroked + unstroked elements).
3. Use the `-medium` variants for each color from the new style guide as the text + arrow colors for _just_ the default select when color modifiers are applied. Also use the `-medium` color variant for the arrow in the stroked selects, but use the default form color for the text color.

adrianababakanian committed:
Use double box-shadow on focus for stroked selects

adrianababakanian committed:
Double box-shadow for default stroked select

samanpwbb reviewed (changes requested):
@adrianababakanian this is looking _really_ good. Almost ready to merge. I see two tiny issues:

Default stroked tab state doesn't show double border:
<img width="337" alt="Screen Shot 2021-04-15 at 9 52 39 AM" src="https://user-images.githubusercontent.com/108094/114907852-5b923900-9dd0-11eb-8879-30d5b631e0b7.png">

Then, after seeing this work and thinking through your notes about differences between stroked/unstroked selects, and thinking through how we use selects throughout our systems now, here's my recommendation: 

- Model unstroked select styling after _links_: leave default and active coloring as you have it now, and _add_ hover states. This also more closely models how Studio now styles unstroked selects 
- Model stroked select styling after _text inputs_: use default text coloring always, and only apply color variants to border and triangle.

adrianababakanian commented:
Thanks @samanpwbb ! I _think_ I might have pushed up [`bbf9a30` (#992)](https://github.com/mapbox/assembly/pull/992/commits/bbf9a30658a739d4ba087d6210e4867642603cb7), which should resolve

> Default stroked tab state doesn't show double border

, while you had already started reviewing -- let me know if the most recent commit fixes that ☝️ issue! I noticed that in hindsight too.

I'll add those recommendations for unstroked/stroked select styling too.

samanpwbb commented on src/forms.css:
```diff
@@ -262,13 +266,43 @@ because it seems to be more cross-browser consistent with all HTML5 input types
 
+/**
+ * Modify a select component so that it has a border.
+ * Set the border (and arrow) color by adding a `select--{color}` modifier to the `select` element along with the `select--stroke` modifier (`*-dark` colors are not available).
```
@adrianababakanian this documentation should clarify that the select--stroke modifier changes how the color modifier gets interpreted.

adrianababakanian committed:
Styling feedback

adrianababakanian committed:
Update select color modifier docs

adrianababakanian committed:
Fix whitespace

adrianababakanian commented:
@samanpwbb this should be ready for your final review!

> Default stroked tab state doesn't show double border

Per https://github.com/mapbox/assembly/pull/992#issuecomment-820592184, should be good on this point.

> Model unstroked select styling after links: leave default and active coloring as you have it now, and add hover states. This also more closely models how Studio now styles unstroked selects

I made these udpates, but kept the default box-shadow on focus. Links don't use box-shadows -- let me know if you think we should keep or remove them for unstroked selects.

> Model stroked select styling after text inputs: use default text coloring always, and only apply color variants to border and triangle.

✅ 

> this documentation should clarify that the select--stroke modifier changes how the color modifier gets interpreted.

This clarification is discussed in the docs for both the default select group and the `select--stroke` modifier. Open to feedback on restructuring docs to make this clearer!

samanpwbb commented on scripts/build-color-variants.js:
```diff
@@ -208,15 +220,18 @@ function buildColorVariants(variables, config) {
+
+        .select--stroke.select--${color}:focus {
+          box-shadow: inset 0 0 0 1px var(--${darkerShade}), var(--focus-shadow);
+        }
```
We only want to use the double-shadow when `[data-assembly-focus-control='visible']` selector is true, so this should be:

```css
[data-assembly-focus-control='visible'] .select--stroke.select--${color}:focus {
  box-shadow: inset 0 0 0 1px var(--${darkerShade}), var(--focus-shadow);
}
```

samanpwbb reviewed (changes requested):
Noticed one small detail, once that's taken care of this should be good to go.

adrianababakanian committed:
Use [data-assembly-focus-control='visible']

adrianababakanian committed:
Add one more focus-selector

samanpwbb reviewed (approved):
This is great! 

I had one other optional thought when playing around with this: should we remove the default border-radius on non-stroked selects? Just to keep the base style a bit more minimal? We'd then apply it to the stroked variant only. I don't think this change is necessary but if you agree with me, might be worth sneaking in.

adrianababakanian commented:
Thanks @samanpwbb !

>  should we remove the default border-radius on non-stroked selects?

Hm, my initial thought was yes, but since we're applying a box-shadow to unstroked selects on focus, I think it might be nice to keep the default border-radius. That way when a default select is focused, the overall effect would be more consistent with the style guide.
